File: iOS 8 by Tutorials/Objective-C/Chapter 13 - Extensions - Document Providers/Imgvue/Imgvue/S3Kit/BucketInfo.m
/*
 * Copyright (c) 2014 Razeware LLC
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#import "BucketInfo.h"
#import "S3Image.h"

@import UIKit;
@import MobileCoreServices;
#import <AWSS3/AWSS3.h>
#import <AWSRuntime/AWSRuntime.h>


NSString* const ACCESS_KEY_ID = @"<#Access Key ID#>";
NSString* const SECRET_KEY = @"<#Secret Key#>";
NSString* const BUCKET_NAME = @"<#Bucket Name#>";

@interface BucketInfo ()
@property (strong, nonatomic) NSMutableArray* files;
@property (strong, nonatomic) AmazonS3Client* s3;
@end

@implementation BucketInfo

- (instancetype)init {
  self = [super init];
  if (self) {
    _files = [NSMutableArray array];
    _s3 = [[AmazonS3Client alloc] initWithAccessKey:ACCESS_KEY_ID withSecretKey:SECRET_KEY];
    _s3.endpoint = [AmazonEndpoints s3Endpoint:US_EAST_1];
    
  }
  return self;
}

- (NSData*) getObject:(NSString*)key bucket:(NSString*)bucketName {
  S3GetObjectRequest* request = [[S3GetObjectRequest alloc] initWithKey:key withBucket:bucketName];
  S3GetObjectResponse* response = [self.s3 getObject:request];
  return response.body;
}

- (NSData*) load:(NSString*)imageKey {
  return [self getObject:imageKey bucket:BUCKET_NAME];
}

- (NSArray*) listObjects:(NSString*)bucket {
  S3ListObjectsRequest* lor = [[S3ListObjectsRequest alloc] initWithName:bucket];
  S3ListObjectsResponse* resp = [self.s3 listObjects:lor];
  S3ListObjectsResult* res = resp.listObjectsResult;
  return res.objectSummaries;
}

- (void) load:(void(^)())initalLoadBlock thumbnailBlock:(void(^)(NSIndexPath* indexPath)) thumbnailBlock {

  NSArray* summaries = [self listObjects:BUCKET_NAME];

  NSArray* fileArray = [[NSFileManager defaultManager] URLsForDirectory:NSCachesDirectory inDomains:NSUserDomainMask];

  NSMutableArray* images = [NSMutableArray array];
  for (S3ObjectSummary* summary in summaries) {
    S3Image* imWrapper = [[S3Image alloc] initWithSummary:summary];
    [images addObject:imWrapper];
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
      NSData* data = [self getObject:summary.key bucket:BUCKET_NAME];
      if (data) {
        NSURL* url = [(NSURL*)fileArray[0] URLByAppendingPathComponent:summary.key];
        [data writeToURL:url atomically:YES];
        imWrapper.url = url;

        NSIndexPath* indexPath = [NSIndexPath indexPathForRow:[images indexOfObject:imWrapper] inSection:0];
        thumbnailBlock(indexPath);
      }
    });
  }

  self.files = images;
  initalLoadBlock();
}

- (NSUInteger) fileCount {
  return self.files.count;
}

- (S3Image*) objAtIndex:(NSIndexPath*)indexPath {
  return self.files[indexPath.row];
}

- (NSError*) putJPEG:(NSData*)imageData key:(NSString*)imageName bucket:(NSString*)bucketName
{
  S3PutObjectRequest *putObjectRequest = [[S3PutObjectRequest alloc] initWithKey:imageName inBucket:bucketName];
  putObjectRequest.contentType = (id) kUTTypeJPEG;
  putObjectRequest.data = imageData;
  
  S3PutObjectResponse* putObjectResponse = [self.s3 putObject:putObjectRequest];
  NSError* error = putObjectResponse.error;
  
  return error;
}

- (void) uploadImage:(UIImage*)image name:(NSString*)name completion:(void(^)(S3Image* imageObj, NSError* error))completion {
  NSData* imageData = UIImageJPEGRepresentation(image, 0.9);
  NSError* error = [self putJPEG:imageData key:name bucket:BUCKET_NAME];
  
  S3Image* imageObj = nil;
  if (!error) {
    imageObj = [[S3Image alloc] init];
    
    NSArray* fileArray = [[NSFileManager defaultManager] URLsForDirectory:NSCachesDirectory inDomains:NSUserDomainMask];
    NSURL* url = [fileArray[0] URLByAppendingPathComponent:name];
    imageObj.url = url;
  }
  completion(imageObj, error);
}

- (void) uploadData:(NSData*)data name:(NSString*)name completion:(void(^)(NSError* error))completion {
  completion([self putJPEG:data key:name bucket:BUCKET_NAME]);
}

@end

File: iOS 8 by Tutorials/Swift/Chapter 13 - Extensions - Document Providers/Imgvue-Starter/Imgvue/S3Kit/BucketInfo.swift
/*
* Copyright (c) 2014 Razeware LLC
*
* Permission is hereby granted, free of charge, to any person obtaining a copy
* of this software and associated documentation files (the "Software"), to deal
* in the Software without restriction, including without limitation the rights
* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
* copies of the Software, and to permit persons to whom the Software is
* furnished to do so, subject to the following conditions:
*
* The above copyright notice and this permission notice shall be included in
* all copies or substantial portions of the Software.
*
* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
* THE SOFTWARE.
*/

import Foundation
import Swift
import UIKit

let ACCESS_KEY_ID = "Enter Your Access Key ID"
let SECRET_KEY = "Enter Your Secret Key"
let BUCKET_NAME = "Enter Your Bucket Name"

public class BucketInfo {
  var files : [S3Image] = []
  let s3 : S3Wrapper
  
  public init() {
    s3 = S3Wrapper(accessKey: ACCESS_KEY_ID, secretKey: SECRET_KEY)
  }
  
  public func load(imageKey : String) -> NSData? {
    return self.s3.getObject(imageKey, bucket: BUCKET_NAME)
  }
  
  public func load(initalLoadBlock : () -> (), thumbnailBlock : (indexPath : NSIndexPath) -> ()) {
    let keys = s3.listObjects(BUCKET_NAME)
  
    let fileManager = NSFileManager()
    let fileArray = fileManager.URLsForDirectory(.CachesDirectory, inDomains: .UserDomainMask)
  
    var images : [S3Image] = []
  
    for maybeKey : AnyObject in keys {
      let key = maybeKey as String
      let imWrapper = S3Image(key: key)
      images.append(imWrapper)
  
      dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0)) {
        if let data : NSData = self.s3.getObject(key, bucket: BUCKET_NAME) {
          let url = fileArray[0].URLByAppendingPathComponent(key);
          data.writeToURL(url, atomically: true)
          imWrapper.url = url;
  
          let arr : NSArray = images
          let row = arr.indexOfObject(imWrapper)
          let indexPath =  NSIndexPath(forRow: row, inSection: 0)
          thumbnailBlock(indexPath: indexPath)
        }
      }
    }
  
    files = images
    initalLoadBlock()
  }
  
  public func fileCount() -> Int {
    return files.count
  }
  
  public func objAtIndex(indexPath : NSIndexPath) -> S3Image {
    return files[indexPath.row]
  }
  
  public func uploadImage(image: UIImage, name : String, completion:(imageObj:S3Image!, error:NSError!) -> ()) {
    let imageData = UIImageJPEGRepresentation(image, 0.9)
    let error = s3.putJPEG(imageData, key: name, bucket: BUCKET_NAME)
  
    var imageObj : S3Image!
    if error == nil {
      imageObj = S3Image()
  
      let fileManger = NSFileManager.defaultManager()
      let fileArray: NSArray = fileManger.URLsForDirectory(.CachesDirectory, inDomains: .UserDomainMask)
      let filePath = "\(fileArray.lastObject)/\(name)"
    }
    completion(imageObj: imageObj, error: error)
  }
  
  public func uploadData(data: NSData, name: String, completion:(error:NSError!) -> ()) {
    let error = s3.putJPEG(data, key: name, bucket: BUCKET_NAME)
    completion(error: error)
  }
  
}
